lib/updatehub/updatehub.c

```diff
@@ -185,44 +196,45 @@ static bool start_coap_client(void)
 		return false;
 	}
 
+	ret = 1;
+
 	ctx.sock = socket(addr->ai_family, SOCK_DGRAM, protocol);
 	if (ctx.sock < 0) {
 		LOG_ERR("Failed to create UDP socket");
-		return false;
+		goto error;
 	}
 
+	ret = -1;
+
 #if defined(CONFIG_UPDATEHUB_DTLS)
 	if (setsockopt(ctx.sock, SOL_TLS, TLS_SEC_TAG_LIST,
 		       sec_list, sizeof(sec_list)) < 0) {
 		LOG_ERR("Failed to set TLS_TAG option");
-		return false;
+		goto error;
 	}
 
 	if (setsockopt(ctx.sock, SOL_TLS, TLS_PEER_VERIFY, &verify, sizeof(int)) < 0) {
 		LOG_ERR("Failed to set TLS_PEER_VERIFY option");
-		return false;
+		goto error;
 	}
 #endif
 
 	if (connect(ctx.sock, addr->ai_addr, addr->ai_addrlen) < 0) {
 		LOG_ERR("Cannot connect to UDP remote");
-		return false;
+		goto error;
 	}
 
 	prepare_fds();
 
-	return true;
-}
+	ret = 0;
+error:
+	freeaddrinfo(addr);
 
-static void cleanup_connection(void)
-{
-	if (close(ctx.sock) < 0) {
-		LOG_ERR("Could not close the socket");
+	if (ret > 0) {
+		cleanup_connection();
```

Comment:
I don't think you need this extra function. `ctx.fds` is set only in `prepare_fds` that is not called in the error path.

Reply:
This is not a new extra function. I moved it up on code because is an internal without prototype. Since we added a for loop based on nfds and proper initialize nfds we can call cleanup_connection. This method will report an error if close is accessed without a valid socket. I think it is better keep only one way to close and clean the socket in this case.